openfl/component/director/director.py

```diff
@@ -0,0 +1,169 @@
+            else:
+                raise Exception('Bad request')
+
+        logger.info(f'New experiment {request.name} for collaborators {request.collaborator_names}')
+        # TODO: save to file
+        self.experiment_data[request.name] = npbytes
+        tensor_dict = None
+        if request.model_proto:
+            tensor_dict, _ = deconstruct_model_proto(request.model_proto, NoCompressionPipeline())
+
+        # TODO: add a logic with many experiments
+        for col_name in request.collaborator_names:
+            queue = self.col_exp_queues[col_name]
+            await queue.put(request.name)
+        self.create_workspace(request.name, npbytes)
+        self._run_aggregator(tensor_dict, request.name)
+        # loop = asyncio.get_event_loop()
+        # fut = loop.run_in_executor(self.executor, self._run_aggregator, tensor_dict, request.name)
+        # await fut
+        # future = self.executor.submit(self._run_aggregator, tensor_dict, request.name)
+        # self.aggregator_task = future
+
+        return director_pb2.Response(accepted=True)
+
+    async def GetExperimentData(self, request, context):
+        # experiment_data = preparations_pb2.ExperimentData()
+        # with open(experiment_name + '.zip', 'rb') as content_file:
+        #     content = content_file.read()
+        #     # TODO: add size filling
+        #     # TODO: add experiment name field
+        #     # TODO: rename npbytes to data
+        content = self.experiment_data.get(request.experiment_name, b'')
+        logger.info(f'Content length: {len(content)}')
+        max_buffer_size = (2 * 1024 * 1024)
+
+        for i in range(0, len(content), max_buffer_size):
+            chunk = content[i:i + max_buffer_size]
+            logger.info(f'Send {len(chunk)} bytes')
+            yield director_pb2.ExperimentData(size=len(chunk), npbytes=chunk)
+
+    # Await prob
+    async def WaitExperiment(self, request_iterator, context):
+        logger.info('Request WaitExperiment has got!')
+        async for msg in request_iterator:
+            logger.info(msg)
+        queue = self.col_exp_queues[msg.collaborator_name]
+        experiment_name = await queue.get()
+        logger.info(f'Experiment {experiment_name} was prepared')
+
+        yield director_pb2.WaitExperimentResponse(experiment_name=experiment_name)
+
+    async def GetShardsInfo(self, request, context):
+        logger.info('Request GetShardsInfo has got!')
+        resp = director_pb2.ShardInfo(
+            sample_shape=self.sample_shape,
+            target_shape=self.target_shape
+        )
+        return resp
+
+    async def GetRegisterdShards(self, request, context):
+        logger.info('Request GetRegisterdShards has got!')
+        resp = director_pb2.GetRegisterdShardsResponse(
+            shard_info=self.shard_registry
+        )
+        return resp
+
+    @staticmethod
+    def create_workspace(experiment_name, npbytes):
```

Comment:
We go through the workspace importing procedure in three places now: here, in envoy's grpc client, and in `workspace import` cli command.
We could describe the procedure in one place and just import it so no code is repeated.

Reply:
We also install python requirements during a workspace import. If we install requirements inside long-living services that means we use the same environment (even the same interpreter session).

Reply:
Could you please create an issue?